uc-crux-llvm/src/UCCrux/LLVM/Constraints.hs

```diff
@@ -0,0 +1,386 @@
+    _relationalConstraints :: [RelationalConstraint m argTypes]
+  }
+
+argConstraints :: Simple Lens (Constraints m argTypes) (Ctx.Assignment (ConstrainedShape m) argTypes)
+argConstraints = lens _argConstraints (\s v -> s {_argConstraints = v})
+
+globalConstraints :: Simple Lens (Constraints m globalTypes) (Map L.Symbol (Some (ConstrainedShape m)))
+globalConstraints = lens _globalConstraints (\s v -> s {_globalConstraints = v})
+
+relationalConstraints :: Simple Lens (Constraints m argTypes) [RelationalConstraint m argTypes]
+relationalConstraints = lens _relationalConstraints (\s v -> s {_relationalConstraints = v})
+
+emptyArgConstraints ::
+  Ctx.Assignment (FullTypeRepr m) argTypes ->
+  Ctx.Assignment (ConstrainedShape m) argTypes
+emptyArgConstraints argTypes =
+  fmapFC
+    ( \argType ->
+        ConstrainedShape
+          (fmapFC (\_ -> Compose []) $ Shape.minimal argType)
+    )
+    argTypes
+
+emptyConstraints ::
+  Ctx.Assignment (FullTypeRepr m) argTypes ->
+  Constraints m argTypes
+emptyConstraints argTypes =
+  Constraints
+    { _argConstraints = emptyArgConstraints argTypes,
+      _globalConstraints = Map.empty,
+      _relationalConstraints = []
+    }
+
+_oneArgumentConstraint ::
+  Ctx.Assignment (FullTypeRepr m) argTypes ->
+  Ctx.Index argTypes inTy ->
+  ConstrainedShape m inTy ->
+  Constraints m argTypes
+_oneArgumentConstraint argTypes idx shape =
+  emptyConstraints argTypes & argConstraints . ixF' idx .~ shape
+
+ppConstraints :: Constraints m argTypes -> Doc Void
+ppConstraints (Constraints args globCs relCs) =
+  PP.vsep
+    ( catMaybes
+        [ if Ctx.sizeInt (Ctx.size args) == 0
+            then Nothing
+            else
+              Just $
+                nestSep
+                  ( PP.pretty "Arguments:" :
+                    toListFC
+                      (Shape.ppShape ppConstraints' . getConstrainedShape)
+                      args
+                  ),
+          if Map.size globCs == 0
+            then Nothing
+            else
+              Just $
+                nestSep
+                  [ PP.pretty "Globals: TODO"
+                  ],
+          if null relCs
+            then Nothing
+            else
+              Just $
+                nestSep
+                  [ PP.pretty "Relational constraints: TODO"
+                  ]
+        ]
+    )
+  where
+    ppConstraints' (Compose constraints) = PP.vsep (map ppConstraint constraints)
+    nestSep = PP.nest 2 . PP.vsep
+
+isEmpty :: Constraints m argTypes -> Bool
+isEmpty (Constraints args globs rels) =
+  allFC (Shape.isMinimal (null . getCompose) . getConstrainedShape) args
+    && Map.null globs
+    && null rels
+
+--------------------------------------------------------------------------------
+
+-- * 'ConstrainedShape'
+
+-- | A specification of the shape of a value and constraints on it.
+newtype ConstrainedShape m ft = ConstrainedShape
+  {getConstrainedShape :: Shape m (Compose [] (Constraint m)) ft}
```

Comment:
It's pipefitting: Shape takes something of kind `FullType -> Type` in that position. I don't think it can be avoided except by hiding it behind it's own newtype.

I'll add a comment, but in short this is a collection of all the constraints we've collected on a value, in the same shape as the value itself. It will be traversed during Setup, where each list of constraints will be replaced with a RegValue that conforms to them. I'll point to that in the comment, too.

Reply:
Added commentary in https://github.com/GaloisInc/crucible/pull/673/commits/6e2cfe7a85e32d8e04b83f5833597b8bcf0fa5e3, let me know if that helps.